autonomica-api/app/services/semrush_client.py

```
@@ -0,0 +1,583 @@
+                return parsed_data
+                
+            except httpx.HTTPStatusError as e:
+                if e.response.status_code >= 500:
+                    # Server error, retry with exponential backoff
+                    delay = min(self.base_delay * (2 ** attempt), self.max_delay)
+                    logger.warning(f"SEMrush API server error: {e}. Retrying in {delay} seconds")
+                    await asyncio.sleep(delay)
+                    continue
+                else:
+                    # Client error, don't retry
+                    raise SEMrushAPIError(f"SEMrush API error: {e}")
```

Ruff (0.12.2)

Within an `except` clause, raise exceptions with `raise ... from err` or `raise ... from None` to distinguish them from errors in exception handling

(B904)